modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Sequence.scala

```diff
@@ -256,6 +256,12 @@ object Sequence {
       Sequence.Zipper.zipper(q).map(Zipper(_, SequenceState.Idle))
         .getOrElse(Final(Sequence.empty(q.id, q.metadata), SequenceState.Idle))
 
+    implicit val eq: Equal[State] = Equal.equal {
```

Comment:
Doing it pairwise is an ok way to do it, but I would say `Equal[Either[zipper fields, final fields]].contramap { ...` where the *fields* need equal instances. You would not normally define `Equal` instances for `Zipper` and `Final` cases because we treat them as data constructors, not data types.

Reply:
How do you go about defining equal for more complex ADTs. like

```
    sealed trait SequenceState
    case object Completed         extends SequenceState
    case object Idle              extends SequenceState
    case object Stopped           extends SequenceState
    final case class Running(userStop: Boolean, internalStop: Boolean) extends SequenceState
```

Nested `Either`?

Reply:
In that case you know that `equalA` will work because the only data members are `Boolean`s.

But yeah you could mechanically write it like

```scala
Equal[Either[Unit, Either[Unit, Either[Unit, (Boolean, Boolean)]]]].contramap {
  case Completed     => Left(())
  case Idle          => Right(Left(())
  case Stopped       => Right(Right(Left(()))
  case Running(a, b) => Right(Right(Right(a, b)))
}
```

Which you could simplify to `Option` since `Option[A] ~ Either[Unit, A]

```scala
Equal[Option[Option[Option[(Boolean, Boolean)]]]].contramap {
  case Completed     => None
  case Idle          => Some(None)
  case Stopped       => Some(Some(None))
  case Running(a, b) => Some(Some(Some(a, b)))
}
```

Or you could use a tuple with  a tag and options of the fields, like 

```scala
Equal[(Int, Option[(Boolea, Boolean)].contramap {
  case Completed     => (1, None)
  case Idle          => (2, None)
  case Stopped       => (3, None)
  case Running(a, b) => (4, Some((a, b)))
}
```

I'm not sure what the best strategy is in general … [kittens](https://github.com/milessabin/kittens) can derive these mechanically but I haven't tried it.

Reply:
I have to say I'm starting to miss the benefit of using stuff like `Eq` if for a very simple ADT you need such a complex construct.
Also, I've read cats documentation and many examples and this is the first time I see the need for `contramap`.

Reply:
Well you can also write it pairwise, which might be the best approach here. You don't *need* to use `contramap`, but you can. Some constructs like `Order` benefit more from this.

Reply:
I guess it would be good being able to create `Equal` from the components `Equal`. i think in `Order` you can do something like that.